node/src/components/block_proposer.rs

```diff
@@ -197,251 +281,118 @@ impl BlockProposer {
-    }
-
     /// Notifies the block proposer that a block has been finalized.
     fn finalized_deploys<I>(&mut self, deploys: I)
     where
         I: IntoIterator<Item = DeployHash>,
     {
+        // TODO: This will ignore deploys that weren't in `pending`. They might be added
+        // later, and then would be proposed as duplicates.
         let deploys: HashMap<_, _> = deploys
             .into_iter()
             .filter_map(|deploy_hash| {
-                self.state
+                self.sets
                     .pending
                     .get(&deploy_hash)
                     .map(|deploy_header| (deploy_hash, deploy_header.clone()))
             })
             .collect();
-        self.state
+        self.sets
             .pending
             .retain(|deploy_hash, _| !deploys.contains_key(deploy_hash));
-        self.state.finalized_deploys.extend(deploys);
+        self.sets.finalized_deploys.extend(deploys);
     }
 
     /// Handles finalization of a block.
     fn handle_finalized_block<I, REv>(
         &mut self,
-        effect_builder: EffectBuilder<REv>,
+        _effect_builder: EffectBuilder<REv>,
         height: BlockHeight,
         deploys: I,
     ) -> Effects<Event>
     where
         I: IntoIterator<Item = DeployHash>,
-        REv: ReactorEventT,
     {
         self.finalized_deploys(deploys);
-        self.state.next_finalized = height + 1;
+        self.sets.next_finalized = height + 1;
 
-        if let Some(requests) = self.state.request_queue.remove(&self.state.next_finalized) {
+        if let Some(requests) = self.request_queue.remove(&self.sets.next_finalized) {
             requests
                 .into_iter()
-                .flat_map(|request| self.get_chainspec(effect_builder, request))
+                .flat_map(|request| {
+                    request
+                        .responder
+                        .respond(self.propose_deploys(
+                            self.deploy_config,
+                            request.current_instant,
+                            request.past_deploys,
+                        ))
+                        .ignore()
+                })
                 .collect()
         } else {
             Effects::new()
         }
     }
 
-    /// Prunes expired deploy information from the BlockProposer, returns the total deploys pruned
-    fn prune(&mut self, current_instant: Timestamp) -> usize {
-        self.state.prune(current_instant)
+    /// Checks if a deploy is valid (for inclusion into the next block).
+    fn is_deploy_valid(
+        &self,
+        deploy: &DeployHeader,
+        block_timestamp: Timestamp,
+        deploy_config: &DeployConfig,
+        past_deploys: &HashSet<DeployHash>,
+    ) -> bool {
+        let all_deps_resolved = || {
+            deploy.dependencies().iter().all(|dep| {
+                past_deploys.contains(dep) || self.sets.finalized_deploys.contains_key(dep)
```

Comment:
I just realized that this now depends on our local clock, too: We might have pruned entries from `finalized_block` because our clock is already a bit ahead of `block_timestamp`, and fail to realize that a dependency is actually satisfied.

Not sure if there's a good solution for that at all. In general, it's probably surprising for the user if deploy dependencies are subject to the TTL.

Reply:
This likely needs to be a ticket as well, but I'd argue it is out of scope for this PR in particular.